### PR TITLE
chore(kit): drain helper extraction + ci pins + buildkit secret

### PR DESCRIPTION
**Pull request by hyochan**
## Summary

- Extract `drainWebhookEventBatches` into a standalone helper (`packages/kit/server/api/v1/webhookStreamDrain.ts`) with unit tests; the SSE live-drain loop in `webhooks.ts` now delegates to it, preserving the same advance/abort semantics so cohort and iteration-limit edges are testable in isolation.
- Drop the redundant `by_project_and_received_and_creation` Convex index. Convex auto-appends `_creationTime` to every index, so `by_project_and_received` already serves both `webhookEventsSince` and `latestWebhookEventsSince` queries.
- Add a real-HTTP SSE integration test for `connectWebhookStream` in `packages/gql` using `Bun.serve` plus a `fetch`-based `EventSource` shim.
- Pass `VITE_KIT_MIXPANEL_TOKEN` as a BuildKit secret instead of an `ARG`/`ENV` pair (`Dockerfile`, `deploy-kit.yml`, `scripts/deploy-prod.sh`) to silence Docker's TOKEN-named warning while still inlining the public token into the SPA bundle.
- Bump `actions/checkout@v4` → `v6`, `docker/setup-buildx-action@v3` → `v4`, and `docker/build-push-action@v6` → `v7` across CI workflows.

## Test plan

- [x] `packages/kit` lint, typecheck, tests (337 passed), smoke server probe — passed locally via pre-push gate
- [x] `packages/gql` `webhook-client.sse.test.ts` runs against a real Bun-served SSE endpoint — 1 pass / 3 expect() calls
- [x] `deploy-kit.yml` deploy step passes Mixpanel token via `--build-secret`; Dockerfile mount uses `required=false` so the PR-level build (which does not pass the secret) still succeeds
- [x] No stale action pins remain (grep for `actions/checkout@`, `setup-buildx-action@`, `build-push-action@` shows only `@v6`, `@v4`, `@v7`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes / Reliability**
  * Improved webhook live-tail draining to handle saturated same-timestamp cohorts, avoid stalled drains, and ensure failed deliveries remain retryable.

* **Tests**
  * Added coverage for additional webhook drain edge cases (saturated cohorts, deduplication, write-failure retry behavior).

* **Chores**
  * Updated CI/CD workflows and Docker tooling to newer releases.
  * Hardened production deploys to pass sensitive tokens securely while minimizing cache churn.

### CODE REVIEW

File: packages/kit/convex/_generated/api.d.ts [hunk 1]
```diff
@@ -58,6 +58,7 @@ import type * as subscriptions_internal from "../subscriptions/internal.js";
 import type * as subscriptions_monthlyMicros from "../subscriptions/monthlyMicros.js";
 import type * as subscriptions_mutation from "../subscriptions/mutation.js";
 import type * as subscriptions_query from "../subscriptions/query.js";
+import type * as subscriptions_selectLatest from "../subscriptions/selectLatest.js";
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/convex/_generated/api.d.ts:61]: The implementation file already exists at `packages/kit/convex/subscriptions/selectLatest.ts` — it landed in #124. The only change here is regenerating `api.d.ts` so the typed Convex client picks up the existing module (you can verify with `git log --oneline -- packages/kit/convex/subscriptions/selectLatest.ts`).

File: packages/gql/src/webhook-client.sse.test.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    }
+  }
+}
+
+describe("connectWebhookStream SSE integration", () => {
+  it("receives typed SSE frames from an actual HTTP stream", async () => {
+    const encoder = new TextEncoder();
+    const server = Bun.serve({
+      port: 0,
+      fetch: () =>
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/gql/src/webhook-client.sse.test.ts]: Fixed in 267f0ed. Added a `test` script to `packages/gql/package.json` (`bun test src`) and a "Run tests" step to the `test-gql` job in `ci.yml`. The new SSE integration test plus the existing `webhook-client.test.ts` and `kit-api.test.ts` files now run on PRs whenever the `gql` filter trips.

File: packages/kit/Dockerfile [hunk 3]
```diff
@@ -44,11 +46,11 @@ ENV VITE_KIT_CONVEX_URL=${VITE_KIT_CONVEX_URL}
...
 ARG VITE_KIT_SENTRY_DSN
 ENV VITE_KIT_SENTRY_DSN=${VITE_KIT_SENTRY_DSN}
-# Mixpanel project token — also public by design (exposed in the SPA
-# bundle to identify users client-side). Empty value disables init.
-ARG VITE_KIT_MIXPANEL_TOKEN
-ENV VITE_KIT_MIXPANEL_TOKEN=${VITE_KIT_MIXPANEL_TOKEN}
-RUN bun run build:all
+# Mixpanel project token is public by design and baked into the SPA when set,
+# but pass it via BuildKit secret so Docker does not flag TOKEN-named ARG/ENV.
+RUN --mount=type=secret,id=VITE_KIT_MIXPANEL_TOKEN,required=false \
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/Dockerfile:61]: Fixed in 267f0ed. Added a `VITE_KIT_MIXPANEL_TOKEN_HASH` build-arg purely as a cache buster — the Dockerfile RUN line now sees a non-secret 16-char sha256 prefix of the token, so a token rotation invalidates the cached `bun run build:all` layer. `deploy-kit.yml` and `scripts/deploy-prod.sh` compute the hash and pass it alongside the BuildKit secret.

File: packages/kit/server/api/v1/webhookStreamDrain.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      hitIterationLimit = true;
+      options.onIterationLimit?.({ iterations, cursor: { ...cursor } });
+      break;
+    }
+    iterations += 1;
+
+    const batch = await options.loadBatch({ ...cursor, limit });
+    if (!batch.length) {
+      break;
+    }
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/server/api/v1/webhookStreamDrain.ts:121]: Fixed in 267f0ed. Ported the saturated-cohort fallback from the original Phase 1 drain into the helper: when the next batch comes back empty AND the previous iteration delivered events at `cursor.sinceMs`, the helper bumps `sinceMs` past the millisecond, resets `afterCreationTime`, and retries before declaring drain complete. Added a unit test (`steps past a saturated millisecond cohort hidden by the query cap`) covering the exact failure mode.

**Comment by hyochan** [comment]
@coderabbitai review

File: packages/kit/server/api/v1/webhookStreamDrain.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    cursor,
+    delivered,
+    iterations,
+    hitIterationLimit,
+  };
+}
+
+function advanceCursor(
+  cursor: WebhookStreamCursor,
+  event: Record<string, unknown>,
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/server/api/v1/webhookStreamDrain.ts]: Fixed in 2af6471. Replaced `Record<string, unknown>` with a `WebhookStreamEvent` type that names the three fields the helper actually reads (`id`, `receivedAt`, `_creationTime`) while keeping an index signature so the SSE writer can pass through additional fields like `type` and the payload body without fighting the type system.

File: packages/kit/scripts/deploy-prod.sh [hunk 6]
```diff
@@ -56,16 +56,23 @@ esac
...
+  BUILD_FLAGS+=(--build-arg "VITE_KIT_SENTRY_DSN=$VITE_KIT_SENTRY_DSN")
 fi
 if [ -n "${VITE_KIT_MIXPANEL_TOKEN:-}" ]; then
-  BUILD_ARGS+=(--build-arg "VITE_KIT_MIXPANEL_TOKEN=$VITE_KIT_MIXPANEL_TOKEN")
+  # Public SPA config, passed as a BuildKit secret to avoid Docker's
+  # TOKEN-named ARG/ENV warning while still baking it into the bundle.
+  # The hash arg is purely a cache buster — BuildKit secret values do
+  # not participate in the layer cache key, so without it a token
+  # rotation would happily reuse the prior `bun run build:all` layer.
+  MIXPANEL_TOKEN_HASH=$(printf '%s' "$VITE_KIT_MIXPANEL_TOKEN" | shasum -a 256 | cut -c1-16)
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/scripts/deploy-prod.sh]: Fixed in 2af6471. deploy-prod.sh now prefers `sha256sum` (the typical Linux choice) and falls back to `shasum -a 256` (macOS). It errors out with a clear message if neither is on PATH so a deployer never silently ships without the cache-bust arg.

**Comment by hyochan** [comment]
@coderabbitai review

File: packages/kit/server/api/v1/webhookStreamDrain.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+      await options.writeEvent(event, id);
+      delivered += 1;
+      if (typeof event.receivedAt === "number") {
+        lastDeliveredReceivedAt = event.receivedAt;
+      }
+    }
+
+    if (!advanced || batch.length < limit) {
+      break;
+    }
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/server/api/v1/webhookStreamDrain.ts:169]: The premature-stop scenario doesn't actually fire — `advanced` is sticky across the inner `for` loop, not reset per-event. The loop iterates the entire batch and `advanced` flips to `true` on **any** event whose cursor moves forward. So a single malformed first event with the next 499 valid still ends the iteration with `advanced = true` and the outer loop keeps draining.

The only path that hits `!advanced` is when **no** event in a full `limit`-sized batch advanced the cursor — which means either every event was malformed (then re-querying would return the same untyped batch forever, and breaking is the correct safety behavior) or every event was at-or-before the cursor (already walked, same outcome). Both cases would otherwise be infinite loops, so `!advanced` breaking is intentional, not premature.

The fact that `WebhookStreamEvent` widens to `unknown` for the three cursor fields is also deliberate: the helper is meant to be defensive against whatever the `loadBatch` callback returns, since callers pass the raw Convex query result. The runtime `typeof` checks in `advanceCursor` cover that contract.

File: packages/kit/Dockerfile [hunk 8]
```diff
@@ -48,6 +48,11 @@ ARG VITE_KIT_SENTRY_DSN
...
 # Mixpanel project token is public by design and baked into the SPA when set,
 # but pass it via BuildKit secret so Docker does not flag TOKEN-named ARG/ENV.
+# BuildKit secret values are not part of the layer cache key, so a non-secret
+# fingerprint of the token is passed as a build arg purely to bust the
+# `bun run build:all` cache when the token rotates — without it, GHA / Fly's
+# remote builder would happily reuse the prior layer and ship a stale token.
+ARG VITE_KIT_MIXPANEL_TOKEN_HASH=
 RUN --mount=type=secret,id=VITE_KIT_MIXPANEL_TOKEN,required=false \
     VITE_KIT_MIXPANEL_TOKEN="$(cat /run/secrets/VITE_KIT_MIXPANEL_TOKEN 2>/dev/null || true)" \
     bun run build:all
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/Dockerfile:64]: Fixed in 19af666. Expanded `VITE_KIT_MIXPANEL_TOKEN_HASH` inside the RUN command (`VITE_KIT_MIXPANEL_TOKEN_HASH="${VITE_KIT_MIXPANEL_TOKEN_HASH}"` prefix) so BuildKit folds it into the layer cache key — you were right that the bare `ARG` was being silently ignored, which would have left the cached `bun run build:all` layer happy to ship a stale token after rotation.

File: packages/kit/server/api/v1/webhookStreamDrain.ts [hunk 9]
```diff
@@ -71,6 +101,9 @@ export async function drainWebhookEventBatches(
       options.seen.add(id);
       await options.writeEvent(event, id);
       delivered += 1;
+      if (typeof event.receivedAt === "number") {
+        lastDeliveredReceivedAt = event.receivedAt;
+      }
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/server/api/v1/webhookStreamDrain.ts]: Fixed in 19af666. Moved `options.seen.add(id)` to **after** the `await options.writeEvent(...)` call. A thrown writer now leaves the event eligible for retry on the next drain pass. Loop is sequential (`for...of` with `await`), so no in-batch race where the same id could be dispatched twice. Added a unit test (`leaves a failed event eligible for retry...`) verifying the retry semantics.

File: packages/kit/Dockerfile [hunk 10]
```diff
@@ -48,6 +48,11 @@ ARG VITE_KIT_SENTRY_DSN
 ENV VITE_KIT_SENTRY_DSN=${VITE_KIT_SENTRY_DSN}
 # Mixpanel project token is public by design and baked into the SPA when set,
 # but pass it via BuildKit secret so Docker does not flag TOKEN-named ARG/ENV.
+# BuildKit secret values are not part of the layer cache key, so a non-secret
+# fingerprint of the token is passed as a build arg purely to bust the
+# `bun run build:all` cache when the token rotates — without it, GHA / Fly's
+# remote builder would happily reuse the prior layer and ship a stale token.
+ARG VITE_KIT_MIXPANEL_TOKEN_HASH=
 RUN --mount=type=secret,id=VITE_KIT_MIXPANEL_TOKEN,required=false \
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/Dockerfile:61]: Fixed in 19af666. The bare `ARG` wasn't part of the cache key, as you flagged — now expanded inside the RUN command (`VITE_KIT_MIXPANEL_TOKEN_HASH="${VITE_KIT_MIXPANEL_TOKEN_HASH}"`) so BuildKit folds it in. A token rotation now invalidates the cached `bun run build:all` layer.

File: packages/kit/server/api/v1/webhookStreamDrain.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/server/api/v1/webhookStreamDrain.ts]: Fixed in 19af666. Renamed and re-wired the tracker to `lastObservedReceivedAt`, updated unconditionally per event (before the `seen` check), so a full same-millisecond page of dedup'd events still trips the fallback. Added a unit test (`triggers the saturated-cohort fallback even when the page is fully dedup'd`) covering exactly that scenario.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -88,6 +88,10 @@ jobs:
           test -f src/generated/Types.swift || exit 1
           test -f src/generated/types.dart || exit 1
 
+      - name: Run tests
+        working-directory: packages/gql
+        run: bun run test
```
[the remark this answers is not in the document]
Comment (hyochan) [reply .github/workflows/ci.yml:93]: Fixed in 19af666. Switched packages/gql to vitest: added `vitest` as a devDep, changed the `test` script to `vitest run src`, and rewrote the SSE integration test to use Node's `http.createServer` instead of `Bun.serve` so it runs under vitest's Node pool. The CI `Run tests` step now exercises the actual vitest suite, not the bun-test compatibility shim.

**Comment by hyochan** [comment]
@coderabbitai review

File: packages/kit/server/api/v1/webhookStreamDrain.ts [hunk 12]
```diff
@@ -63,13 +95,20 @@ export async function drainWebhookEventBatches(
...
       const id = typeof event.id === "string" ? event.id : null;
       if (!id || options.seen.has(id)) {
         continue;
       }
-      options.seen.add(id);
+      // Add to `seen` *after* writeEvent succeeds so a thrown writer
+      // leaves the event eligible for retry on the next drain pass.
+      // The await keeps the loop sequential, so there's no in-batch
+      // race where the same id could be dispatched twice.
       await options.writeEvent(event, id);
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/server/api/v1/webhookStreamDrain.ts:157]: The sequential `await` is intentional and load-bearing here, not a perf oversight:

1. **SSE ordering**: the only caller (`webhooks.ts`) writes through `stream.writeSSE` to a Hono SSE response. Clients reconnect with `Last-Event-Id` and rely on monotonically-increasing IDs — parallel writes would interleave events from a single batch and break consumer cursor semantics.
2. **Same-batch dedupe**: we just (this PR, prior reviewer suggestion) moved `options.seen.add(id)` to after `writeEvent` so a thrown writer leaves the event retryable. That move is only safe because the loop is sequential — under `Promise.all` two events with the same id could both pass the `seen.has` check before either finishes writing.
3. **Cursor advancement**: `lastObservedReceivedAt` and the cohort fallback rely on the in-order traversal to decide when to bump `cursor.sinceMs`.

Throughput is bounded by the SSE channel's natural ordering anyway, and live drain pages are 500 events at a time. If a particular `writeEvent` is slow that's a backpressure signal we want to honor (the SSE consumer is overwhelmed), not paper over with concurrency.

File: packages/kit/server/api/v1/webhookStreamDrain.ts [hunk 13]
```diff
@@ -53,6 +87,26 @@ export async function drainWebhookEventBatches(
...
+          iterations,
+          cursor: { ...cursor },
+          nextSinceMs: cursor.sinceMs + 1,
+          limit,
+        });
+        cursor.sinceMs += 1;
+        cursor.afterCreationTime = undefined;
+        lastObservedReceivedAt = null;
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (hyochan) [reply packages/kit/server/api/v1/webhookStreamDrain.ts]: Fixed in f61f2e0. Tightened the predicate to require all three signals before firing the fallback:

1. `batch.length === limit` (full page),
2. every event in the page shared one `receivedAt`,
3. that `receivedAt` matched `cursor.sinceMs`.

Tracking happens via `pageReceivedAt` + `pageAllSameReceivedAt` inside the inner loop because comparing each event to `cursor.sinceMs` directly doesn't work — `advanceCursor` bumps the cursor to each event's `receivedAt` mid-iteration, making that equality trivially true on mixed pages.

Added a regression test (`does NOT bump sinceMs when a full page merely ends at the cursor ms but spans multiple receivedAts`) that fills a 5-row page with receivedAts `8000..8003` (last two share `8003`) and asserts `onSaturatedCohortFallback` is NOT invoked and `cursor.sinceMs` stays at `8003`.
